Default qutrit mixed execute

Gabriel-Bottrill opened the pull request:
**Context:**
This is the summation of the qutrit mixed device development, this device will be used to run noisy qutrit circuits. It uses the new device API.

**Description of the Change:**
Adds the execute method to DefaultQutritMixed, also sets device supports backprop derivative to True

**Benefits:**
Allows for noisy qutrit circuits to be executed.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

Gabriel-Bottrill commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -243,5 +276,28 @@ def execute(
         circuits: QuantumTape_or_Batch,
         execution_config: ExecutionConfig = DefaultExecutionConfig,
     ) -> Result_or_ResultBatch:
-        """Stub for execute."""
```
I don't add a docstring as it should be the same as device_api's one.

Gabriel-Bottrill commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -243,5 +276,28 @@ def execute(
     ) -> Result_or_ResultBatch:
-        """Stub for execute."""
-        return None
+        if logger.isEnabledFor(logging.DEBUG):
```
I added code for debugging as I have implemented it throughout the other PRs. If I should leave it out, please let me know.

Gabriel-Bottrill commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -67,6 +73,8 @@ def accepted_sample_measurement(m: qml.measurements.MeasurementProcess) -> bool:
 
 
+@simulator_tracking
+@single_tape_support
```
simulator_tracking and single_tape_support modifiers are added. 

I am not sure if I should add simulator tracking as the code that computes shots is not necessarily true as it doesn't take into account commuting GellMann measurements, instead it counts it naively. This is similar to how the code is run on the device, but may not be the best for an actual device as you may want to be more careful with the samples taken.

If anyone has an opinion on if I should rewrite this, remove it, or leave it as is I would appreciate the input.

mudit2812 replied:
@albi3ro could you help out here? I'm still a bit unfamiliar with how the tracker works.

trbromley replied:
Right, it's a question of whether we are tracking for what is actually happening vs what we would want to happen on an optimized hardware device.

We have a similar question in `default.qubit` and solved it in https://github.com/PennyLaneAI/pennylane/pull/4628, with `executions` being the number of hardware executions and `simulations` being the number of times you would simulate.

So in this case, I guess we are correctly reporting `simulations` but `executions` could be lower?

Overall though, I think we should not block this PR - we can create an issue to follow up later.

Gabriel-Bottrill commented on setup.py:
```diff
@@ -59,6 +59,7 @@
             "null.qubit = pennylane.devices.null_qubit:NullQubit",
             "default.qutrit = pennylane.devices.default_qutrit:DefaultQutrit",
             "default.clifford = pennylane.devices.default_clifford:DefaultClifford",
+            "default.qutrit.mixed = pennylane.devices.default_qutrit_mixed:DefaultQutritMixed",
```
I added the device to the setup.py file as it should be ready to use and it makes it consistent with other devices.

Gabriel-Bottrill commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -12,18 +12,78 @@
+np.random.seed(0)
+
+
+def expected_TRX_circ_expval_values(phi, subspace):
```
These three functions are used as helpers

Gabriel-Bottrill commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -33,8 +93,929 @@ def test_debugger_attribute():
+
+@pytest.mark.parametrize("subspace", [(0, 1), (0, 2)])
+class TestSampleMeasurements:
+    """Tests circuits using sample-based measurements.
```
Like with `test_default_qubit.py` I reuse tests from sampling, simulate, and measurement to make sure functionality doesn't change through device.

glassnotes replied:
This is a lot of duplication, if nothing else, does it make sense to just inherit from that class so that you don't have to duplicate all the static methods? (PL team: is that fine convention-wise? other suggestions here?)

mudit2812 replied:
We don't generally inherit from test classes, so I don't personally mind having this duplication in tests. However, in the future, we can consider collecting the tests into one file and parametrizing over the device. This type of duplication is rampant in the `tests/interfaces` module, where the test files for each interface have the same tests, with the only significant difference being the function used to compute gradients/jacobians, as those are dependent on the interface being used.

Gabriel-Bottrill replied:
Alright I will leave it to for the future

Gabriel-Bottrill reviewed (commented):
I made sure to read through all my docstrings for consistency, hopefully this will be an easier PR to review now.

Also, here are some comments explaining parts of the PR and asking questions I needed clarified. Thank you for reviewing!

Gabriel-Bottrill commented:
Hi @albi3ro, @glassnotes, @mudit2812, and @trbromley. This is it! The execute PR, this PR implements execute and makes the `default.qutrit.mixed` device usable (maybe useless until channels are added). I have added some comments above, hopefully it looks good.

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py:
```diff
@@ -0,0 +1,265 @@
+]
+
+
+@pytest.mark.parametrize("mps, expected_exec, expected_shots", shot_testing_combos)
```
Since these are all related tests, they can be included together in a new class.

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py:
```diff
@@ -0,0 +1,265 @@
+    log_records_expected = [
+        (
+            "pennylane.workflow.qnode",
+            ["Creating QNode(func=<function test_execution_debugging"],
```
```suggestion
            ["Creating QNode(func=<function test_execution_debugging>)"],
```
Not sure about this but there is a missing ) and > here

Gabriel-Bottrill replied:
It's supposed to be like this, the actual message includes the pointer to the function `Creating QNode(func=<function test_execution_debugging.<locals>.circuit at 0x7c2ed9c63c40> ...` Along with some other stuff.

glassnotes commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -12,18 +12,43 @@
     assert DefaultQutritMixed().name == "default.qutrit.mixed"
 
 
+def test_shots():
```
Can group these all into one class that focuses on testing generic device properties

Gabriel-Bottrill replied:
Yes, that's a good idea, thank you!

glassnotes commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -33,8 +58,960 @@ def test_debugger_attribute():
+    @pytest.mark.autograd
+    @pytest.mark.parametrize("subspace", [(0, 1), (0, 2)])
+    def test_autograd_results_and_backprop(self, subspace):
+        """Tests execution and gradients of a basic circuit using torch."""
```
```suggestion
        """Tests execution and gradients of a basic circuit using autograd."""
```

glassnotes commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -33,8 +58,960 @@ def test_debugger_attribute():
+            qml.TShift("b"),
+            qml.ControlledQutritUnitary(
+                qml.TRX.compute_matrix(phi) @ qml.TRX.compute_matrix(phi, subspace=(1, 2)),
+                control_wires=("a", "b", -3),
```
What a nasty operation :smiling_imp:

Gabriel-Bottrill replied:
I was trying to implement something similar to the qubit version here, but the controlled qubit stuff is a bit past the controlled qutrit stuff so I implemented this.

glassnotes reviewed (changes requested):
@Gabriel-Bottrill I can see the light! Thanks for all your work on this.

Some small comments within, two major questions I'm leaving top-level:
 - Is there anything that can be done to reduce duplication of the code/tests?
 - At least from what I can see here, it seems a set of broader integration tests is missing, i.e. tests that do

```python
dev = qml.device("default.qutrit.mixed", wires=3)

@qml.qnode(dev)
...
```

Gabriel-Bottrill commented:
@glassnotes I will look at the duplication again tomorrow

trbromley commented on doc/releases/changelog-dev.md:
```diff
@@ -289,6 +289,9 @@
 * Implement `Shots.bins()` method.
   [(#5476)](https://github.com/PennyLaneAI/pennylane/pull/5476)
 
+* Implemented `execute` on `qml.devices.DefaultQutritMixed` device, `execute` can be used to simulate noisy qutrit based circuits.
```
Thanks, yes it would be great to have a single shiny changelog entry that shows off `default.qutrit.mixed`, with all the composing PRs listed together.

Gabriel-Bottrill replied:
It seems like that was added in a different PR, I moved the message for this PR.

trbromley commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -67,6 +73,8 @@ def accepted_sample_measurement(m: qml.measurements.MeasurementProcess) -> bool:
+@simulator_tracking
+@single_tape_support
 class DefaultQutritMixed(Device):
     """A PennyLane device written in Python and capable of backpropagation derivatives.
```
```suggestion
    """A PennyLane Python-based device for mixed-state qutrit simulation.
```

trbromley reviewed (commented):
Thanks @Gabriel-Bottrill!

mudit2812 commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -172,14 +180,35 @@ def __init__(
+            Bool: Whether or not a derivative can be calculated provided the given information.
+
+        """
+        if execution_config is None or execution_config.gradient_method in {"backprop", "best"}:
+            return circuit is None or not circuit.shots
```
```suggestion
        if execution_config is None or execution_config.gradient_method in {"backprop", "best"}:
            if circuit is None:
                return True
            return not circuit.shots and not any(
                isinstance(m.obs, qml.SparseHamiltonian) for m in circuit.measurements
            )
```

This and a test to go along with it.

mudit2812 replied:
You're right, nevermind!

mudit2812 commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -12,29 +12,1004 @@
+        assert qml.math.allclose(jacobian_1, jacobian_3)
+
+
+@pytest.mark.usefixtures("use_legacy_and_new_opmath")
```
🎉

mudit2812 commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -12,29 +12,1004 @@
+
+        # set a global seed both before initialization of the
+        # device and before execution of the tape
+        np.random.seed(456)
+        dev2 = DefaultQutritMixed(seed=123)
+        np.random.seed(789)
+        result2 = dev2.execute(qs)
```
@albi3ro we want to avoid changing the global seed in tests now right?

Gabriel-Bottrill replied:
I had one idea i can get the state of the rng change the rng then revert to the save state. That or I can set it to none at the end of the test.

Gabriel-Bottrill replied:
I undid that because it shows up in a few tests that change global seed.
These tests are similar to the qubit ones and I checked that they are still the same for the qubit version, do I need to change anything?

mudit2812 replied:
Yeah I just checked the `default.qubit` tests and they do the same thing. No changes needed in that acse. Resolving.

mudit2812 commented on tests/devices/test_default_qutrit_mixed.py:
```diff
@@ -80,9 +1055,7 @@ def test_rng_as_seed(self):
-    """Test that the device behaves correctly when provided with a PRNG key and using the JAX interface"""
-
-    # pylint: disable=too-few-public-methods
+    """Test that the device behaves correctly when provided with a JAX PRNG key."""
```
We recently updated how `PRNGKey`s are used for execution. These tests might have to be updated accordingly.

Gabriel-Bottrill replied:
Would that affect how it is written for this device, do I need to update that?

Gabriel-Bottrill replied:
Oops I didnt see above, my bad I will look into it

Gabriel-Bottrill replied:
I updated it to be like how `default_qubit` tests are now.

Gabriel-Bottrill replied:
The updates were seperate from how QutritMixed sampling happens I tried to make the tests more similar but I have to leave some tests out as it will require changing some of the sampling code which I think should be a separate PR. 

Namely I left out `test_prng_key_multi_tapes`  as it does not use a new PRNG and the `test_get_prng_keys` test as that method has not been added

mudit2812 replied:
Ok, we can cover this in the future in that case.

mudit2812 replied:
Sounds good to me. In that case, we don't need to add the `_prng_seed` argument, as that is used for resetting the PRNGKey before new executions.

Gabriel-Bottrill replied:
Good poing i will remove

Gabriel-Bottrill commented:
Hey @albi3ro, @glassnotes,  @mudit2812, @trbromley. I have made all the requested changes and closed all the conversations, but one. Sorry that my response has been delayed on this one I had to balance a few things.

mudit2812 commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -165,21 +173,44 @@ def __init__(
         super().__init__(wires=wires, shots=shots)
         seed = np.random.randint(0, high=10000000) if seed == "global" else seed
         if qml.math.get_interface(seed) == "jax":
+            self._prng_seed = seed
             self._prng_key = seed
             self._rng = np.random.default_rng(None)
         else:
+            self._prng_seed = None
```
As I said in my other comment, since we're making the PRNGKey changes in the future, we can get rid of these for now.

```suggestion
            self._prng_key = seed
            self._rng = np.random.default_rng(None)
        else:
```

mudit2812 reviewed (approved):
Nice work Gabe! Congrats on getting this in. Just a note: the PRNGKey updates are needed for better support for jax-jit, but we can make it a future improevment.

Gabriel-Bottrill commented:
@glassnotes, I added tests for integration, I ended up catching a small bug (for if you want to measure more wires than are in the circuit and fixed that.

mudit2812 commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -50,6 +50,16 @@
 PostprocessingFn = Callable[[ResultBatch], Result_or_ResultBatch]
 
 channels = set()
+observables = {
```
Why is this here if the `observable_stopping_condition` uses `DefaultQutrit.observables`?

Gabriel-Bottrill replied:
My intention was to support them as I realized they were not supported, I removed the call to `qml.DefaultQutrit.observables` too as this device doesn't currently support Identity

Gabriel-Bottrill replied:
I am just working on this now

Gabriel-Bottrill replied:
Added a recursive function to ensure proper observable input

mudit2812 commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -118,23 +121,39 @@ def calculate_probability(
+    # rearrange wires if necessary
+    prob = math.transpose(prob, desired_axes)
+    # flatten and return probabilities
+    return math.reshape(prob, flat_shape)
```
Curious why this change was added.

Gabriel-Bottrill replied:
It only occurs if the number of wires that have operators is less than number of measured wires and there is a gap in the operator wires ie wire 0 and 2 have operators, but 1 doesn't

mudit2812 reviewed (commented):
Tests look good! Just a couple of questions.

mudit2812 commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -43,12 +46,26 @@
+    if obs.name in {"LinearCombination", "Hamiltonian"}:
+        return all(observable_stopping_condition(observable) for observable in obs.terms()[1])
+    if obs.name == "Tensor":
+        return all(observable_stopping_condition(observable) for observable in obs.terms()[1])
```
```suggestion
        return all(observable_stopping_condition(observable) for observable in obs.obs)
```

Gabriel-Bottrill replied:
Fixed

trbromley commented on doc/releases/changelog-dev.md:
```diff
@@ -423,6 +426,7 @@
   (which is not currently compatible with `KerasLayer`), linking to instructions to enable Keras 2.
   [(#5488)](https://github.com/PennyLaneAI/pennylane/pull/5488)
 
+
```
```suggestion
```

trbromley commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -172,14 +193,35 @@ def __init__(
+            circuit (QuantumTape): An optional circuit to check derivatives support for.
+
+        Returns:
+            Bool: Whether or not a derivative can be calculated provided the given information.
```
```suggestion
            bool: Whether or not a derivative can be calculated provided the given information.
```

trbromley commented on pennylane/devices/default_qutrit_mixed.py:
```diff
@@ -201,20 +243,24 @@ def preprocess(
-            TransformProgram, ExecutionConfig: A transform program that when called returns QuantumTapes that the device
-            can natively execute as well as a postprocessing function to be called after execution, and a configuration with
-            unset specifications filled in.
+            TransformProgram, ExecutionConfig: A transform program that when called returns
+                ``QuantumTape`` objects that the device can natively execute, as well as a postprocessing
+                function to be called after execution, and a configuration with unset
+                specifications filled in.
```
```suggestion
            TransformProgram, ExecutionConfig: A transform program that when called returns
            ``QuantumTape`` objects that the device can natively execute, as well as a postprocessing
            function to be called after execution, and a configuration with unset
            specifications filled in.
```
No indent needed, otherwise it shows incorrectly: https://xanaduai-pennylane--5495.com.readthedocs.build/en/5495/code/api/pennylane.devices.default_qutrit_mixed.DefaultQutritMixed.html#pennylane.devices.default_qutrit_mixed.DefaultQutritMixed.preprocess

trbromley reviewed (approved):
Thanks @Gabriel-Bottrill @glassnotes, it's great to have this device added :rocket:

glassnotes reviewed (approved):
Thanks for implementing the final changes @Gabriel-Bottrill !

I think there are a few follow-up things we need to discuss (e.g., more integration tests for the other ML interfaces, and I think there was something else about the PRNG keys?) but the core functionality is there and it should not block completion of the device.

Thanks to everyone on the PL team for helping get this across the finish line!
